### PR TITLE
V 3/wui list select

### PR DESCRIPTION
**Pull request by svenvoskamp**
# Changes

- feat: Add `wui-list-select` component

### CODE REVIEW

File: packages/ui/src/composites/wui-list-select/index.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+@customElement('wui-list-select')
+export class WuiListSelect extends LitElement {
+  public static styles = [globalStyles, styles]
+
+  // -- state & properties ------------------------------------------- //
+  @property({ type: Array }) public walletImages?: IWalletImage[] = []
+
+  @property() public imageSrc? = ''
+
+  @property() public listName = ''
```
Comment (0xAsimetriq) [inline packages/ui/src/composites/wui-list-select/index.ts:19]: Since component already is named with 'list' in it, we don't need to repeat that in props i.e. this can be just `name`

File: packages/ui/src/composites/wui-list-select/index.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+  @property() public listStatus?: string
```
Comment (0xAsimetriq) [inline packages/ui/src/composites/wui-list-select/index.ts:21]: same as above, can be just `status`. As we only have a handful of statuses, perhaps we can make this more strict instead of `string` i.e specific status like installed | recent etc.

File: packages/ui/src/composites/wui-list-select/index.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+  @property({ type: Boolean }) public showAllWallets = false
```
Comment (0xAsimetriq) [inline packages/ui/src/composites/wui-list-select/index.ts:23]: There is no need to for this as we can check if `walletImages` is undefined i.e. further down below:

```
if (this.showAllWallets && this.walletImages) {
 throw new Error('Either showAllWallets or walletImages can be defined, not both')
}

// ...

{this.showAllWallets && <wui-all-wallet....}
{this.imageSrc && <wui-wallet-imag...}
```

File: packages/ui/src/composites/wui-list-select/index.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  // -- render ------------------------------------------------------- //
+  public render() {
+    const getWalletComponent = () => {
```
Comment (0xAsimetriq) [inline packages/ui/src/composites/wui-list-select/index.ts:27]: If you adjust as per previous comment, no need to use if/else here and return undefined

File: packages/ui/src/composites/wui-list-select/index.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      return undefined
+    }
+
+    const status =
+      this.listStatus &&
+      html` <wui-text variant="micro-700" color="fg-300">${this.listStatus}</wui-text> `
+
+    return html`
+      <button>
+        <div>
```
Comment (0xAsimetriq) [inline packages/ui/src/composites/wui-list-select/index.ts:49]: Can we avoid using div here, by making `<wui-text>` for name (as these can be the longest) occupy all remaining available space in flex, thus pushing status label to the end?

**Review by 0xAsimetriq — CHANGES_REQUESTED** [review]
Could you re-ping me for another review once your previous pr for text changes is merged and applied, I believe it is also included here.

File: packages/ui/src/composites/wui-list-select/index.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  @property() public name = ''
+
+  @property() public status?: 'installed' | 'qr code' | 'recent'
+
+  @property({ type: Boolean }) public showAllWallets = false
+
+  // -- render ------------------------------------------------------- //
+  public render() {
+    const visual =
```
Comment (0xAsimetriq) [inline packages/ui/src/composites/wui-list-select/index.ts:27]: When you have sub-components like this, best create separate methods like:

```ts
private visualTemplate() {
  // ...
  return html`...`
}
```

And then use them inside `render()`. Makes things easier to follow and understand when scanning through code after some time when returning to it.

Comment (0xAsimetriq) [reply packages/ui/src/composites/wui-list-select/index.ts:27]: Further down below, you have a relatively complex ternary i.e. `? ... : ... &&`. Those are also hard to follow sometimes, I'd suggest you take out all logic for rendering those into new template functions I suggested above and basically check for all conditions to match before you return `html` or instead just return `null`

This will simplify main render to just

```
// ...
${this.templateAllWallets()}
${this.templateWalletImage()}
// ..
```

File: packages/ui/src/composites/wui-list-select/styles.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import { css } from 'lit'
+
+export default css`
+  button {
+    display: flex;
+    column-gap: var(--wui-spacing-s);
+    align-items: center;
+    width: 336px;
```
Comment (0xAsimetriq) [inline packages/ui/src/composites/wui-list-select/styles.ts:8]: If I understand correctly idea for this component is to occupy full width of its parent in the modal right? So setting static width like this should be avoided.

**Review by 0xAsimetriq — CHANGES_REQUESTED** [review]

File: packages/ui/src/composites/wui-list-select/index.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      <button>
+        ${this.templateAllWallets()} ${this.templateWalletImage()}
+        <wui-text variant="paragraph-500" color="fg-100">${this.name}</wui-text>
+        ${this.templateStatus()}
+      </button>
+    `
+  }
+
+  private templateAllWallets() {
+    if (this.showAllWallets && this.walletImages) {
```
Comment (0xAsimetriq) [inline packages/ui/src/composites/wui-list-select/index.ts:37]: We need to account for less that 4 walletImages being passed here. There is some logic in V2 to use images that were given and replace missing ones with wallet placeholder graphics.

**Review by 0xAsimetriq — CHANGES_REQUESTED** [review]

**Comment by svenvoskamp** [comment]
- Fallback logic added
- Disabled state added

File: apps/gallery/stories/composites/wui-list-select.stories.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import { walletImagesOptions } from '../../utils/PresetUtils'
+
+type Component = Meta<WuiListSelect>
+
+export default {
+  title: 'Composites/wui-list-select',
+  args: {
+    walletImages: walletImagesOptions,
+    imageSrc:
+      'https://explorer-api.walletconnect.com/w3m/v1/getWalletImage/7a33d7f1-3d12-4b5c-f3ee-5cd83cb1b500?projectId=a8d876c6f91c3748db621583fad358f1',
```
Comment (0xAsimetriq) [inline apps/gallery/stories/composites/wui-list-select.stories.ts:15]:
```suggestion
    imageSrc: walletImagesOptions[0].src
```

File: packages/ui/src/composites/wui-all-wallets-image/index.ts [hunk 10]
```diff
@@ -1,31 +1,39 @@
...
 
 @customElement('wui-all-wallets-image')
 export class WuiAllWalletsImage extends LitElement {
   public static styles = [globalStyles, styles]
 
   // -- state & properties ------------------------------------------- //
-  @property({ type: Array }) public walletImages: WalletImage[] = []
+  @property({ type: Array }) public walletImages: IWalletImage[] = []
+
+  @state() public totalImages = 4
```
Comment (0xAsimetriq) [inline packages/ui/src/composites/wui-all-wallets-image/index.ts:15]: We should try to avoid using state in ui package, you can define constants outside of class via simple `const TOTAL_IMAGES = 4` or if you need to latter update this value, you can make it a private class one i.e. `private totalImages = 4`.

State should be used only when we need to trigger component's re-render from it self, which in case of ui we want to avoid, because we want scaffold package to control this logic through `@properties`

File: packages/ui/src/composites/wui-all-wallets-image/index.ts [hunk 11]
```diff
@@ -1,31 +1,39 @@
...
-          size="inherit"
-          src=${src}
-          walletName=${walletName}
-        ></wui-wallet-image>`
-    )
+    return html`${this.templateWalletImages()}`
+  }
+
+  private templateWalletImages() {
+    const renderedImages = this.walletImages
```
Comment (0xAsimetriq) [inline packages/ui/src/composites/wui-all-wallets-image/index.ts:23]: We can simplify this i.e. loop through existing images and add them, then add placeholders for everything else, so

```
${renderedImages.map( w => html`...`)}
${[...Array(TOTAL_IMAGES - renderedImages.length)].map(w => html`...placeholder`)}
```

**Review by 0xAsimetriq — CHANGES_REQUESTED** [review]

**Review by 0xAsimetriq — APPROVED** [review]
